lightly/loss/emp_ssl_loss.py

```diff
@@ -0,0 +1,111 @@
+
+    def forward(self, z_list: List[Tensor]) -> Tensor:
+        num_patch = len(z_list)
+        z_stacked = torch.stack(z_list, dim=0)
+        z_avg = z_stacked.mean(dim=0)
+
+        z_sim = torch.tensor(0.0, device=z_stacked.device, requires_grad=True)
+        for i in range(num_patch):
+            z_sim = z_sim + F.cosine_similarity(z_stacked[i], z_avg, dim=1).mean()
+
+        z_sim = z_sim / num_patch
+
+        return -z_sim
+
+
+class TotalCodingRate(nn.Module):
+    def __init__(self, eps: float = 0.2) -> None:
+        super(TotalCodingRate, self).__init__()
+        self.eps = eps
+
+    def compute_discrimn_loss(self, W: Tensor) -> Tensor:
+        """Discriminative Loss."""
+        d, b = W.shape
+        I = torch.eye(d, device=W.device)
+        scalar = d / (b * self.eps)
+        logdet = torch.logdet(I + scalar * W.matmul(W.T))
+        return logdet / 2.0
+
+    def forward(self, X: Tensor) -> Tensor:
+        return -self.compute_discrimn_loss(X.T)
+
+
+class EMP_SSLLoss(nn.Module):
+    """Implementation of the loss from 'EMP-SSL: Towards
+    Self-Supervised Learning in One Training Epoch' [0].
+    - [0] EMP-SSL, 2023, https://arxiv.org/abs/2304.03977
+
+    Examples:
+        >>> # initialize loss function
+        >>> loss_fn = EMP_SSLLoss(num_patches=100)
+        >>> base_transform = VICRegViewTransform() # As discussed in paper
+        >>> transform_fn = MultiCropTransform(transforms=base_transform, crop_counts=100)
+        >>>
+        >>> # generate the transformed samples
+        >>> samples = transform_fn(image)
+        >>>
+        >>> # feed through encoder head
+        >>> z = torch.cat([model(s) for s in samples])
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(z)
+    """
+
+    def __init__(
+        self,
+        eps: float = 0.2,
+        num_patches: int = 100,
+        tcr_coef: float = 1.0,
+        patch_coef: float = 200.0,
+    ) -> None:
+        """Args as in [0].
+        Args:
+            eps:
+                Total Coding Rate (TCR) epsilon.
+            num_patches:
+                Number of patches to be used.
+            tcr_coef:
+                Coefficient for the TCR loss.
+            patch_coef:
+                Coefficient for the patch loss.
+        """
+        super().__init__()
+        self.similarity_loss = SimilarityLoss()
+        self.tcr = TotalCodingRate(eps=eps)
+        self.num_patches = num_patches
+        self.tcr_coef = tcr_coef
+        self.patch_coef = patch_coef
+
+    def forward(self, z: Tensor) -> Tensor:
+        z_list = z.chunk(self.num_patches, dim=0)
+        contractive_loss = self.similarity_loss(z_list)
+        tcr_loss = torch.tensor(0.0, device=z.device, requires_grad=True)
+        for i in range(self.num_patches):
+            tcr_loss = tcr_loss + self.tcr(z_list[i])
+        tcr_loss = tcr_loss / self.num_patches
+        return Tensor(self.patch_coef * contractive_loss + self.tcr_coef * tcr_loss)
```

Comment:
If there are many views, we usually pass a list of views to the loss (see [DINOLoss](https://github.com/lightly-ai/lightly/blob/534cdf42a0feebe10a917bbdaacc6a533c1104fa/lightly/loss/dino_loss.py#L80-L85) or [VICRegLLoss](https://github.com/lightly-ai/lightly/blob/534cdf42a0feebe10a917bbdaacc6a533c1104fa/lightly/loss/vicregl_loss.py#L98-L104)). I suggest to change the function definition to `def forward(self, z_views: Sequence[Tensor]) -> Tensor`.

And if we go for the approach where the sub-losses are defined as functions the forward method can be rewritten as:
```
    def forward(self, z_views: Sequence[Tensor]) -> Tensor:
        # z has shape (num_views, batch_size, dim)
        z = torch.stack(z_views)
        return tcr_loss(z, eps=self.tcr_eps) + self.inv_coef * invariance_loss(z)
```